File: plugins/plugin-goals/rust/src/actions/create_goal.rs
//! Create goal action

use async_trait::async_trait;
use serde_json::Value;

use super::{ActionContext, GoalAction};
use crate::error::Result;

/// Action to create a new goal
pub struct CreateGoalAction;

#[async_trait]
impl GoalAction for CreateGoalAction {
    fn name(&self) -> &'static str {
        "CREATE_GOAL"
    }

    fn description(&self) -> &'static str {
        "Create a new goal for tracking"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        // Check if there's goal information in the state
        let has_goal_info = context
            .state
            .get("extracted_goal")
            .map(|v| v.get("name").is_some())
            .unwrap_or(false);
        Ok(has_goal_info)
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let extracted = context.state.get("extracted_goal");
        let name = extracted
            .and_then(|e| e.get("name"))
            .and_then(|v| v.as_str())
            .unwrap_or("Unnamed Goal");

        let description = extracted
            .and_then(|e| e.get("description"))
            .and_then(|v| v.as_str());

        Ok(serde_json::json!({
            "action": self.name(),
            "goal": {
                "agent_id": context.agent_id,
                "owner_id": context.entity_id,
                "owner_type": "entity",
                "name": name,
                "description": description
            }
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_with_goal_info() {
        let action = CreateGoalAction;

        let context = ActionContext {
            message: serde_json::json!({}),
            agent_id: "agent-123".to_string(),
            entity_id: "user-456".to_string(),
            room_id: None,
            state: serde_json::json!({
                "extracted_goal": {
                    "name": "Learn Rust"
                }
            }),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_without_goal_info() {
        let action = CreateGoalAction;

        let context = ActionContext {
            message: serde_json::json!({}),
            agent_id: "agent-123".to_string(),
            entity_id: "user-456".to_string(),
            room_id: None,
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_execute() {
        let action = CreateGoalAction;

        let context = ActionContext {
            message: serde_json::json!({}),
            agent_id: "agent-123".to_string(),
            entity_id: "user-456".to_string(),
            room_id: None,
            state: serde_json::json!({
                "extracted_goal": {
                    "name": "Learn Rust",
                    "description": "Master the Rust programming language"
                }
            }),
        };

        let result = action.execute(&context).await.unwrap();
        assert_eq!(result["goal"]["name"], "Learn Rust");
    }
}

File: plugins/plugin-goals/typescript/actions/createGoal.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  composePrompt,
  type Entity,
  formatMessages,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  parseKeyValueXml,
  type State,
  type UUID,
} from "@elizaos/core";
import {
  checkSimilarityTemplate,
  extractGoalTemplate,
} from "../generated/prompts/typescript/prompts.js";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { createGoalDataService } from "../services/goalDataService";

interface GoalInput {
  name: string;
  description?: string;
  ownerType: "agent" | "entity";
}

interface SimilarityCheckResult {
  hasSimilar: boolean;
  similarGoalName?: string;
  confidence: number;
}

async function extractGoalInfo(
  runtime: IAgentRuntime,
  message: Memory,
  state: State
): Promise<GoalInput | null> {
  try {
    const messageHistory = formatMessages({
      messages: (state.data?.messages as Memory[]) || [],
      entities: (state.data?.entities as Entity[]) || [],
    });

    const prompt = composePrompt({
      state: {
        text: message.content.text || "",
        messageHistory,
      },
      template: extractGoalTemplate,
    });

    const result = await runtime.useModel(ModelType.TEXT_LARGE, {
      prompt,
      stopSequences: [],
    });

    logger.debug("Extract goal result:", result);

    // Parse XML from the text results
    const parsedResult = parseKeyValueXml(result);

    if (!parsedResult || !parsedResult.name) {
      logger.error("Failed to extract valid goal information from XML");
      return null;
    }

    return {
      name: String(parsedResult.name),
      description: parsedResult.description ? String(parsedResult.description) : undefined,
      ownerType: (parsedResult.ownerType === "agent" ? "agent" : "entity") as "agent" | "entity",
    };
  } catch (error) {
    logger.error("Error extracting goal information:", error);
    return null;
  }
}

interface ExistingGoal {
  name: string;
  description?: string;
}

async function checkForSimilarGoal(
  runtime: IAgentRuntime,
  newGoal: GoalInput,
  existingGoals: ExistingGoal[]
): Promise<SimilarityCheckResult> {
  try {
    if (existingGoals.length === 0) {
      return { hasSimilar: false, confidence: 0 };
    }

    const existingGoalsText = existingGoals
      .map((goal) => `- ${goal.name}: ${goal.description || "No description"}`)
      .join("\n");

    const prompt = composePrompt({
      state: {
        newGoalName: newGoal.name,
        newGoalDescription: newGoal.description || "No description",
        existingGoals: existingGoalsText,
      },
      template: checkSimilarityTemplate,
    });

    const result = await runtime.useModel(ModelType.TEXT_SMALL, {
      prompt,
      stopSequences: [],
    });

    const parsedResult = parseKeyValueXml(result) as SimilarityCheckResult | null;

    if (!parsedResult) {
      return { hasSimilar: false, confidence: 0 };
    }

    return {
      hasSimilar: String(parsedResult.hasSimilar) === "true",
      similarGoalName: parsedResult.similarGoalName,
      confidence: parseInt(String(parsedResult.confidence || 0), 10),
    };
  } catch (error) {
    logger.error("Error checking for similar goals:", error);
    return { hasSimilar: false, confidence: 0 };
  }
}

const spec = requireActionSpec("CREATE_GOAL");

export const createGoalAction: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,

  validate: async (_runtime: IAgentRuntime, _message: Memory): Promise<boolean> => {
    return true;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    try {
      const currentState = state || (await runtime.composeState(message, ["GOALS"]));
      const goalInfo = await extractGoalInfo(runtime, message, currentState);

      if (!goalInfo) {
        if (callback) {
          await callback({
            text: "I couldn't understand what goal you want to create. Could you please provide a clear goal description?",
            actions: ["CREATE_GOAL_FAILED"],
            source: message.content.source,
          });
        }
        return {
          success: false,
          error: "Could not understand goal description",
        };
      }

      const dataService = createGoalDataService(runtime);
      const ownerId = goalInfo.ownerType === "agent" ? runtime.agentId : (message.entityId as UUID);
      const activeGoalCount = await dataService.countGoals(goalInfo.ownerType, ownerId, false);

      if (activeGoalCount >= 10) {
        if (callback) {
          await callback({
            text: `Cannot add new goal: The ${goalInfo.ownerType === "agent" ? "agent" : "user"} already has 10 active goals, which is the maximum allowed. Please complete or remove some existing goals first.`,
            actions: ["CREATE_GOAL_LIMIT_REACHED"],
            source: message.content.source,
          });
        }
        return { success: false, error: "Goal limit reached" };
      }

      const existingGoals = await dataService.getAllGoalsForOwner(goalInfo.ownerType, ownerId);
      const similarityCheck = await checkForSimilarGoal(runtime, goalInfo, existingGoals);

      if (similarityCheck.hasSimilar && similarityCheck.confidence > 70) {
        if (callback) {
          await callback({
            text: `It looks like there's already a similar goal: "${similarityCheck.similarGoalName}". Are you sure you want to add this as a separate goal?`,
            actions: ["CREATE_GOAL_SIMILAR_EXISTS"],
            source: message.content.source,
          });
        }
        return { success: false, error: "Similar goal exists" };
      }

      const tags = ["GOAL"];
      if (goalInfo.ownerType === "agent") {
        tags.push("agent-goal");
      } else {
        tags.push("entity-goal");
      }

      const metadata: Record<string, unknown> = {
        createdAt: new Date().toISOString(),
      };

      const createdGoalId = await dataService.createGoal({
        agentId: runtime.agentId,
        ownerType: goalInfo.ownerType,
        ownerId,
        name: goalInfo.name,
        description: goalInfo.description || goalInfo.name,
        metadata,
        tags,
      });

      if (!createdGoalId) {
        throw new Error("Failed to create goal");
      }

      let successMessage = `✅ New goal created: "${goalInfo.name}"`;

      if (activeGoalCount >= 4) {
        successMessage += `\n\n⚠️ You now have ${activeGoalCount + 1} active goals. Consider focusing on completing some of these before adding more.`;
      }

      if (callback) {
        await callback({
          text: successMessage,
          actions: ["CREATE_GOAL_SUCCESS"],
          source: message.content.source,
        });
      }
      return { success: true, text: successMessage };
    } catch (error) {
      logger.error("Error in createGoal handler:", error);
      if (callback) {
        await callback({
          text: "I encountered an error while creating your goal. Please try again.",
          actions: ["CREATE_GOAL_FAILED"],
          source: message.content.source,
        });
      }
      return {
        success: false,
        error: error instanceof Error ? error.message : "Unknown error",
      };
    }
  },

  examples: (spec.examples ?? []) as ActionExample[][],
};

export default createGoalAction;
